src/core/connection_pool.c

```diff
@@ -0,0 +1,521 @@
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetInterfaceIndexForLocalAddress(
+    _In_ QUIC_ADDR* LocalAddress,
+    _Out_ uint32_t* InterfaceIndex
+    )
+{
+    CXPLAT_ADAPTER_ADDRESS* Addresses = NULL;
+    uint32_t AddressesCount;
+
+    *InterfaceIndex = 0;
+
+    QUIC_STATUS Status =
+        CxPlatDataPathGetLocalAddresses(
+            MsQuicLib.Datapath,
+            &Addresses,
+            &AddressesCount);
+    if (QUIC_SUCCEEDED(Status)) {
+
+        for (uint32_t i = 0; i < AddressesCount; i++) {
+            if (QuicAddrCompareIp(LocalAddress, &Addresses[i].Address)) {
+                *InterfaceIndex = Addresses[i].InterfaceIndex;
+                break;
+            }
+        }
+
+        if (*InterfaceIndex == 0) {
+            Status = QUIC_STATUS_NOT_FOUND;
+            QuicTraceLogError(
+                ConnPoolLocalAddressNotFound,
+                "[conp] Failed to find local address, 0x%x",
+                Status);
+        }
+
+        CXPLAT_FREE(Addresses, QUIC_POOL_DATAPATH_ADDRESSES);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QUIC_API
+MsQuicConnectionPoolCreate(
+    _In_ QUIC_CONNECTION_POOL_CONFIG* Config,
+    _Out_writes_(Config->NumberOfConnections)
+        HQUIC* ConnectionPool
+    )
+{
+    QUIC_CONNECTION** Connections = (QUIC_CONNECTION**)ConnectionPool;
+    CXPLAT_RSS_CONFIG* RssConfig = NULL;
+    uint32_t* RssProcessors = NULL;
+    uint32_t* ConnectionCounts = NULL;
+    const char* ServerNameCopy = NULL;
+    CXPLAT_TOEPLITZ_HASH ToeplitzHash;
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    uint32_t CreatedConnections = 0;
+
+    if (Config == NULL || ConnectionPool == NULL || Config->Registration == NULL ||
+        Config->Configuration == NULL|| Config->NumberOfConnections == 0 ||
+        Config->Handler == NULL || Config->ServerName == NULL || Config->ServerPort == 0 ||
+        (Config->Family != QUIC_ADDRESS_FAMILY_UNSPEC &&
+            Config->Family != QUIC_ADDRESS_FAMILY_INET &&
+            Config->Family != QUIC_ADDRESS_FAMILY_INET6) ||
+        (((QUIC_CONFIGURATION*)Config->Configuration)->SecurityConfig == NULL) ||
+        (Config->CibirIds != NULL && Config->CibirIdLength == 0) ||
+        (Config->CibirIds == NULL && Config->CibirIdLength != 0)) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolInvalidParam,
+            "[conp] Invalid parameter, 0x%x",
+            Status);
+        return Status;
+    }
+
+    const size_t ServerNameLength = strnlen(Config->ServerName, QUIC_MAX_SNI_LENGTH + 1);
+    if (ServerNameLength == QUIC_MAX_SNI_LENGTH + 1) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolServerNameTooLong,
+            "[conp] ServerName too long (%llu bytes)",
+            ServerNameLength);
+        return Status;
+    }
+
+    QuicTraceEvent(
```

Comment:
FYI, in general, for other functions in api.c we are logging api enter/exit even for INVALID_PARAMETER cases.

Reply:
Right, but I can't dereference an untrusted pointer without validating it isn't NULL. So either I log entry after the validation, or I log entry without any parameter, which seems to be the requirement for this scenario.